src/pull-mode/index.js

```diff
@@ -0,0 +1,23 @@
+/**
+ * Utility for libraries from the `Lodash`.
+ *
+ * @ignore
+ */
+import { filter, last, sortBy } from 'lodash';
+
+/**
+ * Determines which element has the highest occurrence (mode) in the given array.
+ *
+ * @function
+ * @since      1.1.0
+ * @param  	   {Array}     input    The array to process.
+ * @return 	   {string}             Mode element.
+ * @example
+ *
+ * pullMode( ['pear', 'apple', 'orange', 'apple'] );
+ *
+ * // => string "apple"
+ */
+const pullMode = ( input ) => last( sortBy( input, ( a, b ) => filter( input, ( v ) => v === a ).length - filter( input, ( v ) => v === b ).length ) );
```

Comment:
I'm a little concerned about the efficiency of this function. `sortBy` is already a non-linear operation and we are calling the callback of`filter` n * (n-1) times for all of the n items. For large input arrays, this might become an issue.

I would thus propose a different function:

```JS
const pullMode = ( input ) => {
	if ( ! input || ! input.length ) {
		return null;
	}

	let maxItem = head( input );
	const itemCounts = {};

	forEach( input, ( item ) => {
		itemCounts[ item ] = itemCounts[ item ] ? itemCounts[ item ] + 1 : 1;
		if ( itemCounts[ item ] > itemCounts[ maxItem ] ) {
			maxItem = item;
		}
	} );

	return maxItem;
}
```

Reply:
As @gooklani mentioned, this is almost a replica of a function deep inside [Gutenberg core](https://github.com/WordPress/gutenberg/blob/f1f0cd7adceeecce546641dafdb2a1d0844f2bb2/packages/components/src/box-control/utils.js#L41-L56) which we have no access to reuse, and I have added this to our package since I would need this in a few extensions that I’m working on.

The primary reason I have decided to replace the vanilla JavaScript `sort` function with `sortBy` is that the method from Lodash performs a stable sort. It preserves the original sort order of equal elements.

Reply:
Would you thus be in favour of keeping the function body from Gutenberg? IMO this is a good opportunity to improve the computational complexity of this operation. We also have the necessary unit tests in place to confirm that the two functions work as expected.

Reply:
Sure, Would there be any performance benefit with doing so? My understanding is that in most cases, using Lodash would be much more optimal and error-free than vanilla JavaScript prototypes. Have you seen any significant difference in benchmarking these two proposals with Chrome profiler or other universally available tools?

Reply:
> 

Absolutely. Using [benchmarkjs](https://benchmarkjs.com/) and benchmarking both functions on my local using a small array, my proposal is approx. 6x faster. Using a large array, the linear function is more than 10x as fast.

Input `[ 'a', 'a', 'a', 'b', 'b', 'c', 'c' ]`:
```
pullMode first x 357,327 ops/sec ±1.33% (91 runs sampled)
pullMode linear x 2,435,302 ops/sec ±0.57% (94 runs sampled)
```

Input `[ 'a', 'a', 'a', 'b', 'b', 'c', 'c', 'a', 'a', 'a', 'b', 'b', 'c', 'c', 'a', 'a', 'a', 'b', 'b', 'c', 'c', 'a', 'a', 'a', 'b', 'b', 'c', 'c', 'a', 'a', 'a', 'b', 'b', 'c', 'c' ]`:
```
pullMode first x 26,089 ops/sec ±1.16% (88 runs sampled)
pullMode linear x 437,529 ops/sec ±0.91% (89 runs sampled)
```

Reply:
You would be able to verify these numbers by installing [benchmarkjs](https://benchmarkjs.com/) and running the following node script:

```JS
var lodash = require('lodash');
var Benchmark = require('benchmark');

function pullMode( input ) {
	return lodash.last( lodash.sortBy( input, ( a, b ) => lodash.filter( input, ( v ) => v === a ).length - lodash.filter( input, ( v ) => v === b ).length ) );
}

function pullModeLinear( input ) {
	if ( ! input || ! input.length ) {
		return null;
	}

	let maxItem = lodash.head( input );
	const itemCounts = {};

	lodash.forEach( input, ( item ) => {
		itemCounts[ item ] = itemCounts[ item ] ? itemCounts[ item ] + 1 : 1;
		if ( itemCounts[ item ] > itemCounts[ maxItem ] ) {
			maxItem = item;
		}
	} );

	return maxItem;
}


var suite = new Benchmark.Suite;
suite.add('pullMode first', function() {
	pullMode( [ 'a', 'a', 'a', 'b', 'b', 'c', 'c' ] );
})
	.add('pullMode linear', function() {
		pullModeLinear( [ 'a', 'a', 'a', 'b', 'b', 'c', 'c' ] );
	})
	// add listeners
	.on('cycle', function(event) {
		console.log(String(event.target));
	})
	.on('complete', function() {
		console.log('Fastest is ' + this.filter('fastest').map('name'));
	})
	// run async
	.run({ 'async': true });
```